Add mechanism for storing warnings and deprecations outside of AnsibleModule

samdoran opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Create a mechanism for storing warnings in a global place, outside of an AnsibleModule object.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request


##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`lib/ansible/module_utils/basic.py`
`lib/ansible/module_utils/common/warnings.py`

samdoran committed:
Add mechanism for warnings outside of AnsibleModule

bcoca commented on lib/ansible/module_utils/common/warnings.py:
```diff
@@ -0,0 +1,22 @@
+global_warnings = []
+
+
+class AnsibleWarning(Warning):
```
we also have a 'debug' facility that works the same, should we just add it to this PR?

samdoran replied:
Might not be a bad idea. I'm not familiar with that mechanism.

samdoran replied:
After working through this more, I don't think moving `debug` out would make sense with this change. That is more a flag rather than a container for holding values. We could handle that in other PR maybe, but it seems out of context for this change.

bcoca replied:
current debug just pushes to local syslog, but with this functioning the same as 'warnings' and return accumulated messages back to controller

samdoran committed:
Move warn() and deprecate() methods out

samdoran committed:
Use global_warnings and global_deprications

samdoran committed:
Don't mess with debug

samdoran committed:
Use imported global values safely

samdoran committed:
Mark the globals as private

samdoran committed:
Add unit tests for new warn and deprecate functions

Andersson007 commented on test/units/module_utils/common/warnings/test_deprecate.py:
```diff
@@ -0,0 +1,50 @@
+        1,
+        True,
+        [1],
+        {'k1': 'v1'},
```
```suggestion
        b'bytestr',
        {u'k1': u'v1'},
```

Andersson007 commented on test/units/module_utils/common/warnings/test_warn.py:
```diff
@@ -0,0 +1,45 @@
+        1,
+        True,
+        [1],
+        {'k1': 'v1'},
```
```suggestion
        b'bytestr',
        {u'k1': u'v1'},
```

Andersson007 commented on lib/ansible/module_utils/common/warnings.py:
```diff
@@ -0,0 +1,25 @@
+        raise TypeError("warn requires a string not a %s" % type(warning))
+
+
+def deprecate(msg, version=None):
```
Maybe we should check version kwarg too?
```suggestion
def deprecate(msg, version=None):
```

samdoran replied:
That is covered in `test_deprecate_with_version()`

Andersson007 commented on test/units/module_utils/common/warnings/test_deprecate.py:
```diff
@@ -0,0 +1,50 @@
+
+
+@pytest.mark.parametrize(
+    'test_case',
```
Regarding that I suggested above
```suggestion
    'test_case,version',
```

Andersson007 commented on test/units/module_utils/common/warnings/test_deprecate.py:
```diff
@@ -0,0 +1,50 @@
+
+def test_deprecate_message_only():
+    deprecate('Deprecation message')
+    warnings._global_deprecations == [{'msg': 'Deprecation message', 'version': None}]
```
```suggestion
    assert warnings._global_deprecations == [{'msg': 'Deprecation message', 'version': None}]
```

samdoran replied:
That is covered in `test_deprecate_with_version()`

samdoran replied:
Good catch!

bcoca commented on lib/ansible/module_utils/basic.py:
```diff
@@ -2018,8 +2013,8 @@ def _return_formatted(self, kwargs):
 
-        if self._warnings:
-            kwargs['warnings'] = self._warnings
+        if warnings._global_warnings:
```
if you are externally accessing it, it should not be a private var, or you might want to consider a read only method instead

samdoran replied:
That may not be a bad idea. Ideally updating these values could only be done via `warn()` and `deprecate()`.

samdoran committed:
Change tests based on feedback

samdoran committed:
Fix unit test and change reference to self._warnings

Store the warning message in the new global variable

samdoran committed:
Fix unit tests for deprecations

samdoran committed:
Rename vars so they're not marked private

samdoran committed:
Also store debug messages globally

I'm not sure how much use this has since we don't do anything with them outside of the `log()` method.

Rename filed to "notice" since it contains more than warning-related functions.

samdoran commented on lib/ansible/module_utils/common/notice.py:
```diff
@@ -0,0 +1,33 @@
+
+global_warnings = []
+global_deprecations = []
+global_debugs = []
```
I'm not crazy about this name. It seemed to fit with the other names though `global_debug_messages` seems less janky.

samdoran committed:
Revert "Also store debug messages globally"

This reverts commit 63db773dae0ecc4f96a921df61344ec19d92c7bb.

sivel commented on lib/ansible/module_utils/basic.py:
```diff
@@ -1426,7 +1419,7 @@ def _handle_aliases(self, spec=None, param=None, option_prefix=''):
         alias_results, self._legal_inputs = handle_aliases(spec, param, alias_warnings=alias_warnings)
         for option, alias in alias_warnings:
-            self._warnings.append('Both option %s and its alias %s are set.' % (option_prefix + option, option_prefix + alias))
+            warnings.global_warnings.append('Both option %s and its alias %s are set.' % (option_prefix + option, option_prefix + alias))
```
Why are you not using the helper functions here?

samdoran replied:
I think I missed this one. I had a deep discussion with @abadger about the nuances of importing globals. I'm pretty sure I need to use the helper function and not modify the var directly.

samdoran committed:
Use helper functions when adding warnings and deprecations

samdoran committed:
Add changelog

samdoran committed:
Fix unit tests

acozine commented on changelogs/fragments/add-global-warnings-container.yaml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - add mechaninsm for storing warnings and deprecations globally and not attached to an ``AnsibleModule`` object (https://github.com/ansible/ansible/pull/58993)
```
Typo:
```suggestion
  - add mechanism for storing warnings and deprecations globally and not attached to an ``AnsibleModule`` object (https://github.com/ansible/ansible/pull/58993)
```

samdoran replied:
Doh! Thanks for catching that.

samdoran committed:
Fix typo in changelog

abadger commented on lib/ansible/module_utils/common/warnings.py:
```diff
@@ -0,0 +1,25 @@
+from ansible.module_utils.six import string_types
+
+global_warnings = []
+global_deprecations = []
```
tldr; These need to be marked private.  Create a pair of simple accessor functions to give a read-only view of the data.  (Example implementation at the bottom)

These need to be marked private for three reasons: they're only implementation details of the framework, it's easy to imagine changes to the implementation (for instance, they could become a set.  Or you could start storing structured data like line numbers and calling functions inside of there), and we don't want people to modify them outside of the framework.

I see that you changed from private to public already as a result of a review comment.  The review comment does note something that's generally something to watch for but the solution of making it public is wrong.  Two things to note:

* In general, things outside of a framework touching private variables inside of the framework is wrong.  However, AnsibleModule is "special".  It is something that we are going to deprecate in the *long* term as we move things (via PRs like this) outside of it because AnsibleModule is an overgrown god object which violates the separation of concerns.  However, that combination of a god object which we are in the process of pulling apart does inately mean that it could be valid for it to touch private values inside of the pieces pulled from it.  For instance, if we decided that some of the existing API should only be supported in the deprecated AnsibleModule and implementing that API required using those private variables, then that would be a valid use case.

* However, in this case, we'll probably find that once we move the fail_json()/exit_json() style code outside of AnsibleModule, that code will need to still access the accumulated warnings and deprecations.  So what we want to do is provide some new API that can give non-legacy code access to the values stored in here without giving them the ability to modify it outside of the framework.  So the right implementation is probably accessor functions.  (Accessor is just a fancy way to say, I need a function to get the value of the data but it implies that the value cannot be modified unless there's a corresponding mutator function as well)

I think this would be a proper implementation of an accessor:
``` python
_global_warnings = []

def get_warning_messages():
    """Return a tuple of the warning messages accumulated over this run of the program"""
    return tuple(_global_warnings)
```

* The `get_` prefix to the function name is one thing that clues the user in that this is an accessor.
* Transforming the list into a tuple tells the user that they cannot modify the warnings via this function
* Using `_messages` in the function name allows you the leeway to write a new function that returns additional information should that be added in the future.

samdoran replied:
So this is just a convention rather than a protection mechanism? I don't see anything that is preventing direct modification or use of the variables marked as private. How would someone know not to modify/reference `warnings._global_warnings` directly and instead go through the accessor/mutator functions?

abadger replied:
Right.  By naming with a leading underscore, you are marking it as private by convention but python doesn't build in anything to prevent someone from accessing it anyways.

samdoran committed:
Change variable to private and create accessor functions

This lays the foundation for future functions being moved outside of AnsibleModule
that need an interface to these items without modifying them.

- Update existing unit tests
- Add new unit tests for new accessor functions

abadger commented on lib/ansible/module_utils/basic.py:
```diff
@@ -2019,8 +2020,8 @@ def _return_formatted(self, kwargs):
-        if warnings.global_warnings:
-            kwargs['warnings'] = warnings.global_warnings
+        if get_warning_messages():
+            kwargs['warnings'] = get_warning_messages()
```
Whenever I call a function twice like this (first to check if the value meets a condition and then to make use of that value) I like to turn it into calling the function once and storing the value.  That way the cost of calling the function and any logic the function has to perform are only done once.

``` python

warnings = get_warning_messages()
if warnings:
    kwargs['warnings'] = warnings
```

In this case, we're only saving the cost of calling the function and converting a list into a tuple (and only doing it once per module) so it shouldn't be a big overhead.  But it is nice to do as not everyone will know that the implementation of `get_warning_messages()` is so lightweight (and who knows, it might become a more costly operation in the future)

samdoran replied:
Thanks for this feedback. I had this same question when I was writing it. The cost of calling isn't too high, so is it worth calling once and saving the result? I was hoping you had thoughts. 🧙🏽‍♂️

> ...not everyone will know that the implementation of get_warning_messages() is so lightweight (and who knows, it might become a more costly operation in the future)

This is a very good point. _I_ know it's lightweight, but future readers might not. And it may change in the future. Thank you for the guidance.

abadger commented:
Just one comment (which applies to two places)  Once those two places are fixed, this looks good to me.

samdoran committed:
Call once and store the value

abadger commented:
Looks good to me!

rebuild_merge

felixfontein commented:
This PR mangles the warning message generated when `removed_in_version` is used in argument_spec:
```
[DEPRECATION WARNING]: msg. This feature will be removed in version version. Deprecation warnings can be disabled by setting deprecation_warnings=False in 
```
is now shown instead of
```
[DEPRECATION WARNING]: Param 'x' is deprecated. See the module docs for more information. This feature will be removed in version 2.11. Deprecation warnings 
can be disabled by setting deprecation_warnings=False in ansible.cfg.
```
I've included a fix in #66918.
